test-data/unit/check-literal-string.test

```diff
@@ -0,0 +1,104 @@
+-- LiteralString tests
+-- See https://peps.python.org/pep-0675/
+
+[case testLiteralStringInference]
+from typing_extensions import LiteralString
+
+x: LiteralString = 'a'
+raw_str: str = x   # Ok, can be narrowed
+
+some_str: str
+y: LiteralString = some_str  # E: Incompatible types in assignment (expression has type "str", variable has type "LiteralString")
+
+z: LiteralString = 1  # E: Incompatible types in assignment (expression has type "int", variable has type "LiteralString")
+[builtins fixtures/literal_string.pyi]
+
+
+[case testLiteralTypeAndLiteralString]
+from typing_extensions import LiteralString, Literal
+
+l1: Literal['a']
+l2: Literal[1]
+
+ls1: LiteralString = l1
+ls2: LiteralString = l2  # E: Incompatible types in assignment (expression has type "Literal[1]", variable has type "LiteralString")
+
+ls3: LiteralString
+l3: Literal['a'] = ls3  # E: Incompatible types in assignment (expression has type "LiteralString", variable has type "Literal['a']")
+
+def expects_literal_string(x: LiteralString): ...
+def expects_literal_a(x: Literal['a']): ...
+
+expects_literal_string(l1)
+expects_literal_string(ls1)
+
+expects_literal_a(l1)
+expects_literal_a(ls1)  # E: Argument 1 to "expects_literal_a" has incompatible type "LiteralString"; expected "Literal['a']"
+[builtins fixtures/literal_string.pyi]
+
+
+[case testLiteralStringFallbackToString]
+from typing_extensions import LiteralString
+def expects_literal_string(x: LiteralString): ...
+
+x: LiteralString
+expects_literal_string(x.format(1))  # E: Argument 1 to "expects_literal_string" has incompatible type "str"; expected "LiteralString"
+[builtins fixtures/literal_string.pyi]
+
+
+-- TODO: this is not supported yet
+-- All cases here must pass
+-- But, we need literal type math for this
+[case testLiteralStringTypeMath-skip]
+from typing_extensions import LiteralString
+def expects_literal_string(x: LiteralString): ...
+
+expects_literal_string('a')
+expects_literal_string('a' + 'b')
+expects_literal_string('a' * 2)
```

Comment:
Why do we need to care about this at all? Isn't this just a question of defining these correctly in typeshed? For example:
```python
@overload
def __add__(self: LiteralStr, other: LiteralStr) -> LiteralStr: ...
@overload
def __add__(self, other: str) -> str: ...
```
It seems to be much easier to agree on this once and for all rather than special-casing all these in all type-checkers.

Reply:
Good question! I think that we should have both.
Typeshed will store the base implementation for it.

But, mypy can go further with full literal type math.
Or we can just use the typeshed if ltm is never merged / implemented.

Reply:
OK, I see, but as I said below, we need to be sure "typeshed way" will work well. Having `"a" + "b"` accepted where `Literal["ab"]` is expected is lower priority IMO.